src/snowflake/cli/api/sql_contract.py

```diff
@@ -0,0 +1,134 @@
+        self.role = role
+        super().__init__(ObjectType.ROLE, role)
+
+
+class SQLService:
+    _sql_executor = get_sql_executor()
+
+    # TODO: Extract common to a _use_object_optional
+    @contextmanager
+    def _use_warehouse_optional(self, new_wh: str | None):
+        """
+        Switches to a different warehouse for a while, then switches back.
+        This is a no-op if the requested warehouse is already active or if no warehouse is passed in.
+        If there is no default warehouse in the account, it will throw an error.
+        """
+        if new_wh is None:
+            yield
+        wh_result = self._sql_executor.execute_query(
+            f"select current_warehouse()"
+        ).fetchone()
+
+        # If user has an assigned default warehouse, prev_wh will contain a value even if the warehouse is suspended.
+        try:
+            prev_wh = wh_result[0]
+        except:
+            prev_wh = None
+        # new_wh is not None, and should already be a valid identifier, no additional check is performed here.
+        is_different_wh = new_wh != prev_wh
+        if is_different_wh:
+            self._sql_executor.log_debug(f"Using warehouse: {new_wh}")
+
+            try:
+                self._sql_executor.execute_query(f"use warehouse {new_wh}")
+            except ProgrammingError as err:
+                if err.errno == DOES_NOT_EXIST_OR_CANNOT_BE_PERFORMED:
+                    raise CouldNotUseObjectError(ObjectType.WAREHOUSE, new_wh) from err
+                else:
+                    raise UnknownSQLError(f"Failed to use warehouse {new_wh}") from err
+            except:
+                raise UnknownSQLError(f"Failed to use warehouse {new_wh}")
+
+        try:
+            yield
+
+        finally:
+            if is_different_wh and prev_wh is not None:
+                self._sql_executor.log_debug(f"Switching back to warehouse:{prev_wh}")
+                self._sql_executor.execute_query(f"use warehouse {prev_wh}")
+
+    @contextmanager
+    def _use_role_optional(self, new_role: str | None):
+        """
+        Switches to a different role for a while, then switches back.
+        This is a no-op if the requested role is already active or if no role is passed in.
+        """
+        if new_role is None:
+            yield
+
+        prev_role = self._sql_executor.current_role()
+        is_different_role = (
+            new_role is not None and new_role.lower() != prev_role.lower()
+        )
+        if is_different_role:
+            self._sql_executor.log_debug(f"Assuming different role: {new_role}")
+            try:
+                self._sql_executor.execute_query(f"use role {new_role}")
+            except ProgrammingError as err:
+                if err.errno == DOES_NOT_EXIST_OR_CANNOT_BE_PERFORMED:
+                    raise CannotUseRoleError(new_role) from err
+                else:
+                    raise UnknownSQLError(f"Failed to use role {new_role}") from err
```

Comment:
```suggestion
                raise UnknownSQLError(f"Failed to use role {new_role}") from err
```
no need for `else` after `raise`